Track file metadata from transfer METS when sending to transfer backlog

mistydemeo opened the pull request:
In artefactual/archivematica#153, the transfer METS has been augmented with additional metadata about the transfer and files, providing the information that ES also carries about them. This improves the File model to add the appropriate fields to store that metadata, then provides new endpoints to

a) List all files in a package, with their metadata;
b) Query for file metadata using the same names as in the transferfile ES index.

The latter can be used by Archivematica to look up specific records via absolute values from the SS, instead of querying the ES index.

Commit:
File: add additional fields

These fields allow additional metadata to be tracked on files in a
package; in particular, it allows the Storage Service to understand
and return information on files in backlogged transfers.

Commit:
PackageResource: allow fetching metadata for files

Commit:
Package: track metadata for backlogged files when ingested

Hwesta commented on storage_service/locations/migrations/0007_auto__add_field_file_package__add_field_file_source_package__add_field.py:
```diff
@@ -0,0 +1,248 @@
+# -*- coding: utf-8 -*-
```
This should have a more useful name.  Migrations can be generated with names with `manage.py schemamigration <app> --auto <name>`; it will prefix the number automatically.

mistydemeo replied:
Updated to something descriptive.

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -769,3 +771,57 @@ def _attempt_package_request_event(self, package, request_info, event_type, even
+        return http.HttpResponse(status=200, content=json.dumps(response),
+            content_type='application/json')
+
+    def file_data(self, request, **kwargs):
```
Docstrings for these? I know we don't have them on other functions, but they sure are nice.

mistydemeo replied:
Added!

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -559,6 +560,130 @@ def compress_package(self, algorithm, extract_path=None):
+            prefix = self.full_path
+
+        mets_path = os.path.join(prefix, *relative_path)
+        if not os.path.exists(mets_path):
```
`if not os.path.isfile(mets_path):`?

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -559,6 +560,130 @@ def compress_package(self, algorithm, extract_path=None):
+
+        doc = etree.parse(mets_path)
+
+        mets_ns = utils.NSMAP['mets']
```
Unnecessary, since mets_ns is only used once?

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -559,6 +560,130 @@ def compress_package(self, algorithm, extract_path=None):
+
+        mets_ns = utils.NSMAP['mets']
+        namespaces = {'m': mets_ns}
+        mets = doc.xpath('/m:mets', namespaces=namespaces)
```
`doc.find` would return one element or None, allowing you to avoid mets[0].

mistydemeo replied:
`doc.find` doesn't implement the full XPATH syntax (while `doc.xpath`) does, so not all of these queries work with `.find`. I'll change the ones that do though.

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -559,6 +560,130 @@ def compress_package(self, algorithm, extract_path=None):
+        if accession_ids:
+            accession_id = accession_ids[0].text
+        else:
+            accession_id = ""
```
`header.findtext` may be useful here.

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -589,6 +714,12 @@ def backlog_transfer(self, origin_location, origin_path):
+        try:
+            self._index_file_data_from_transfer_mets(prefix=os.path.join(dest_space.staging_path, self.current_path))  # create File entries for every file in the transfer
+        except StorageException as e:
+            LOGGER.warning("Transfer METS data could not be read: %s", str(e))
```
If this happens, has it failed to be stored in backlog correctly?  Should we return a failed status code?  It can't be arranged if the Files table isn't populated, and if it can't be removed from backlog it's kinda useless.

mistydemeo replied:
Yeah, I'm a little torn on that one. You can still fetch it, just not necessarily in a useful or straightforward way.

Hwesta replied:
We discussed this with Justin - what was the decision?  To fail loudly?

mistydemeo commented:
Pushed some changes in response to feedback.

Hwesta commented:
SIP arrange fails with files that differ between sanitize and unsanitized, eg evelyn's photo.jpg vs evelyn_s_photo.jpg.  Dunno if fix belongs here or artefactual/archivematica#153

mistydemeo commented:
Ugh, I didn't consider that. Even though the `structMap` gets updated the `fileSec` doesn't, and the `fileSec` will have pre-sanitization names.

mistydemeo commented:
Updated: now uses the `amdSec` from the transfer METS to look up changed names. I've tested this with Unicode and SJIS-named files that get sanitized during the transfer.

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -559,6 +561,135 @@ def compress_package(self, algorithm, extract_path=None):
+            raise StorageException("<metsHdr> element not found in METS file!")
+
+        try:
+            creation_date = header.attrib['CREATEDATE']
```
`header` is a list from xpath, so this fails.  `header = doc.find('m:metsHdr', namespaces=namespaces)` seems to work.

mistydemeo replied:
Fixed.

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -559,6 +561,135 @@ def compress_package(self, algorithm, extract_path=None):
+        doc = etree.parse(mets_path)
+
+        namespaces = {'m': utils.NSMAP['mets'],
+                      'p': utils.NSMAP['premis']}
```
utils.NSMAP has all the namespaces I've used in this project.

mistydemeo replied:
Yes, I was just creating an nsmap with shorter names - the longer xpath queries were getting quite long, so I found it easier to read with brief prefixes.

Hwesta commented:
Tests for the new endpoints?

mistydemeo commented:
Added tests for the new endpoints and for the METS parsing, and rebased those into their respective commits.

Commit:
Package: add endpoints to remove, add files to packages

These support a new archivematica-devtools script to reindex files
from the ES index.

Hwesta commented on storage_service/locations/tests/test_models.py:
```diff
@@ -1,9 +1,34 @@
 from locations import models
 
 
+class TestPackage(TestCase):
```
I had started separating the tests as well (see test_api, test_duracloud, test_locations).  Move this to test_package.py?  Eventually we can rename test_models to test_lockss, sinces that's what it really is...

:tada: Tests! :tada:

mistydemeo replied:
Moved into test_package.py!

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -769,3 +773,176 @@ def _attempt_package_request_event(self, package, request_info, event_type, even
+        return http.HttpNoContent()
+
+    @_custom_endpoint(expected_methods=['post'])
+    def add_files_to_package(self, request, bundle, **kwargs):
```
Rather than having add_files and remove_files, why not have file/<uuid>/files/ that responds to DELETE or PUT? Seems more RESTful.

mistydemeo replied:
Sure, will update.

mistydemeo replied:
Updated!

Hwesta commented on storage_service/locations/migrations/0007_file_add_new_fields.py:
```diff
@@ -0,0 +1,248 @@
+
+        # Adding field 'File.origin'
+        db.add_column(u'locations_file', 'origin',
+                      self.gf('django.db.models.fields.CharField')(default='', unique=False, max_length=36, blank=True),
```
I get a slightly different migration when I generate it. This column is not unique.  Perhaps this file needs to be re-generated?

mistydemeo replied:
Re-generated.

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -589,6 +722,12 @@ def backlog_transfer(self, origin_location, origin_path):
+        try:
+            self.index_file_data_from_transfer_mets(prefix=os.path.join(dest_space.staging_path, self.current_path))  # create File entries for every file in the transfer
+        except StorageException as e:
+            LOGGER.warning("Transfer METS data could not be read: %s", str(e))
```
Is this suppose to raise?

mistydemeo replied:
No - I left this condition here to allow the newer storage service to continue serving older AMs.

Hwesta replied:
Ahhhh, okay. Will those SIPs be arrange-able?

mistydemeo replied:
Within that AM version, yep! After an upgrade to Archivematica 1.4, they'll need to be reindexed in the storage service using the script that's there.

Hwesta replied:
Ah, okay. Makes sense.

Hwesta commented:
:+1:

Commit:
Update CHANGELOG.md for #7714
